ghproxy/ghcache/coalesce.go

```diff
@@ -0,0 +1,108 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ghcache
+
+import (
+	"bufio"
+	"bytes"
+	"net/http"
+	"net/http/httputil"
+	"sync"
+)
+
+// requestCoalescer allows concurrent requests for the same URI to share a
+// single upstream request and response.
+type requestCoalescer struct {
+	sync.Mutex
+	keys map[string]*responseWaiter
+
+	delegate http.RoundTripper
+}
+
+type responseWaiter struct {
+	*sync.Cond
+
+	waiting bool
+	resp    []byte
+	err     error
+}
+
+// RoundTrip coalesces concurrent GET requests for the same URI by blocking
+// the later requests until the first request returns and then sharing the
+// response between all requests.
+//
+// Notes: Deadlock shouldn't be possible because the map lock is always
+// acquired before responseWaiter lock if both locks are to be held and we
+// never hold multiple responseWaiter locks.
+func (r *requestCoalescer) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Only coalesce GET requests
+	if req.Method != http.MethodGet {
+		return r.delegate.RoundTrip(req)
+	}
+
+	key := req.URL.String()
+	r.Lock()
+	waiter, ok := r.keys[key]
+	if ok {
+		// Earlier request in flight. Wait for it's response.
+		defer req.Body.Close() // Since we won't pass the request we must close it.
+		waiter.L.Lock()
+		r.Unlock()
+		waiter.waiting = true
+		waiter.Wait()
+		waiter.L.Unlock()
+		// Earlier request completed.
+
+		if waiter.err != nil {
+			return nil, waiter.err
+		}
+		resp, err := http.ReadResponse(bufio.NewReader(bytes.NewBuffer(waiter.resp)), nil)
+		if err != nil {
+			return nil, err
+		}
+
+		cacheCounter.WithLabelValues("COALESCED").Inc()
+		return resp, nil
+	}
+	// No earlier request in flight (common case).
+	// Register a new responseWaiter and make the request ourself.
+	waiter = &responseWaiter{Cond: sync.NewCond(&sync.Mutex{})}
+	r.keys[key] = waiter
+	r.Unlock()
+
+	resp, err := r.delegate.RoundTrip(req)
+	// Real response received. Remove this responseWaiter from the map THEN
+	// wake any requesters that were waiting on this response.
+	r.Lock()
+	delete(r.keys, key)
+	r.Unlock()
+
+	waiter.L.Lock()
+	if waiter.waiting {
+		if err != nil {
+			waiter.resp, waiter.err = nil, err
+		} else {
+			// Copy the response before releasing to waiter(s).
+			waiter.resp, waiter.err = httputil.DumpResponse(resp, true)
+		}
+		waiter.Broadcast()
+	}
+	waiter.L.Unlock()
+
+	cacheCounter.WithLabelValues(cacheResponseMode(resp.Header)).Inc()
+	return resp, err
```

Comment:
Not sure if we care but in this case you get the response regardless of whether `err == nil` whereas if you get coalesced you get _either_ the response or the error.

Reply:
Fixed.